### PR TITLE
core: prepare introduction of package scopes

### LINKED ISSUE #246: Scope system rework
_Continuously updated_

Tracker for details regarding the current and future form of Bait's scope system.

Changes will be split up into multiple PRs. 
 

### Hierarchy:
```
global scope
+-- file 1
  +-- blocks
    +-- ...
  +-- ...
+-- file 2
+-- ...
```

### Storage:
`table.Table.global_scope &ast.Scope`

### Contents high-level:
- global scope
  - consts
  - globals
  - foreign types
- file scope
  - import aliases
  - functions
- block scope
  - _everything else_
  

### ScopeObject value variants:
```
// normal
Obj{
  name = full.pkg.name
  pkg = full.pkg
}

// builtin
Obj{
  name = name
  pkg = "builtin"
}

// ffi
Obj{
  name = LANG.pkg.name
  pkg = full.pkg
  lang = lang
}
```

### Hierarchy:
```
universe
+-- ffi
+-- pkg 1
  +-- block (file 1)
    +-- blocks
      +-- ...
  +-- block (file 2)
  +-- ...
+-- pkg 2
+-- ...
```

### Storage:
`table.Table.pkg_scopes map[string]&ast.Scope`

### Contents high-level:
- universe
  - primitive types
- FFI
  - all C / JS declarations depending on the backend
- package scope
  - globals
  - constants
  - functions
- block scope (file toplevel)
  - import aliases
- block scope (sub blocks)
  - _everything else_

### ScopeObject value variants:
```
// normal
table.pkg_scopes[full.pkg]Obj{
  name = name
}

// builtin
table.pkg_scopes["builtin"]Obj{
  name = name
}

// ffi
table.pkg_scopes["FFI"]Obj{
  name = pkg.name
  lang = lang // this might not even be needed
}
```

### PR DESCRIPTION
**Pull request by serkonda7**
For details regarding the design please look at #246 (regularly updated).

Part of #246

### CODE REVIEW

File: lib/bait/ast/scope.bt [hunk 1]
```diff
@@ -26,6 +26,29 @@ pub enum ObjectKind {
 	label
 }
 
+pub fun (mut t Table) preregister_scopes(pkg_name string) &Scope {
+	// FFI scopes
+	t.get_pkg_scope("C", 0 as any)
+	t.get_pkg_scope("JS", 0 as any)
```
Comment (StunxFS) [inline lib/bait/ast/scope.bt]: Wouldn't it be better to create just one of the two based on the backend the user is using?

Comment (serkonda7) [reply lib/bait/ast/scope.bt]: You are right. The unused backend's scope would just be empty. I will change it

File: lib/bait/ast/table.bt [hunk 2]
```diff
@@ -3,7 +3,8 @@
 package ast
 
 pub struct Table {
-	pub mut global_scope Scope
+	pub mut global_scope &Scope // TODO entirely replace by scopes
```
Comment (StunxFS) [inline lib/bait/ast/table.bt]: This field could perfectly function as an entry point to all other scopes. As a mere suggestion, you could have the primitive types/pkgs/ffi pkgs and builtin functions (if there are any) be added to this scope.

Comment (serkonda7) [reply lib/bait/ast/table.bt]: Another good find! 
Actually I have already in mind to add a `universe` scope or reintroduce the `global` scope for this purpose.
But primitive types are still resolved differently, so this is for a follow-up change.

Comment (StunxFS) [reply lib/bait/ast/table.bt]: It's a good idea to use `universe` as the root scope ;)

File: lib/bait/ast/scope.bt [hunk 3]
```diff
@@ -27,9 +27,8 @@ pub enum ObjectKind {
 }
 
 pub fun (mut t Table) preregister_scopes(pkg_name string) &Scope {
-	// FFI scopes
-	t.get_pkg_scope("C", 0 as any)
-	t.get_pkg_scope("JS", 0 as any)
+	// FFI is used for either C or JS declarations depending on the used backend
+	t.get_pkg_scope("FFI", 0 as any)
```
Comment (StunxFS) [inline lib/bait/ast/scope.bt:31]: and JS?

**Review by StunxFS — APPROVED** [review]

File: lib/bait/ast/table.bt (hunk 2 above)
Comment (serkonda7) [reply lib/bait/ast/table.bt]: Cannot be confused with `global` keyword

Comment (StunxFS) [reply lib/bait/ast/table.bt]: I suggest not using `global`, use static package-level variables instead.

**Comment by StunxFS** [comment]
```sh
/home/runner/work/bait/bait/bait2.js:384
	return v > MAX_I32 ? -v : v
	           ^

ReferenceError: MAX_I32 is not defined
    at i32 (/home/runner/work/bait/bait/bait2.js:384:13)
```

CI fails because of that 🤔

**Comment by serkonda7** [comment]
@StunxFS wow, you are fast!
But I am aware, it's a change breaking the bootstrap process. The "fix" is merged in #250.
It's a little ugly in terms of development workflow but now I can continue with replacing the global scope.

**Comment by StunxFS** [comment]
And I deleted the comment because the PR had already been merged XD

**Pull request merged**